### ISSUE TITLE
Duplicate request not working

### ISSUE THREAD

**Issue by oskar-fjeld** [a85/POSTMan-Chrome-Extension#5920]
Please read through the [guidelines](https://github.com/postmanlabs/postman-app-support#guidelines-for-reporting-issues) before creating a new issue.

Select Duplicate to duplicate a request and it only creates a new empty request

1. Go to Left Sidebar
2. Click on a Collection
3. Open it
4. Click the three dots next to a request and choose Duplicate

**Expected behavior**
A new request with the data from selected request

**Actual behavior**
A new empty request is created

**App information (please complete the following information):**
 - Postman Version 6.7.3

**Comment by sujayvenaik** [comment]
Hey @oskar-fjeld 
Could you help us with a screen recording of the above situation where you faced this?
Also, can you send us logs from View -> Developer -> View logs?

I tried on v6.7.3 but could not reproduce.

**Comment by oskar-fjeld** [comment]
Hi again

I'll add a couple of screenshots.  There are no new logs to for the day I'm doing this.
![duplicate_context_menu](https://user-images.githubusercontent.com/46679245/52779287-b63c7880-303f-11e9-83fb-0b9dd8ea18ba.png)
![duplicated](https://user-images.githubusercontent.com/46679245/52779288-b63c7880-303f-11e9-979e-db91aa1a4196.png)

**Comment by numaanashraf** [comment]
@oskar-fjeld Were all the changes to the request saved before you had duplicated the request?

**Comment by martyscholes** [comment]
I have a similar issue where duplicating a request produces a different request than that duplicated. I noticed something similar last year where work on one computer (even when saved to the cloud) provided different requests on another computer when cloud sync was in use.

I have a screen movie of the failure in action if you will tell me how to upload it. This comment system does not accept .mov files.

**Comment by martyscholes** [comment]
I am now thinking that this is somehow related to the request tabs at the top of Postman. I can't find the documentation to explain what those tabs are, but they get tied somehow to the requests in the collection. Saving via the save button or the menu does not appear to resolve conflicts between the collection and the tabs at the top. Duplicating a request in a collection appears to duplicate the unresolved conflicting request from one of the tabs at the top. When I tried to close the tabs, I was given dozens of dialog boxes about conflicts. Now that the tabs are all closed, the collections appear to be working properly.

**Comment by lauxjpn** [comment]
Duplicate did not work for me as described.
After explicitly saving the request I wanted to duplicate, it worked as expected.
I think it would be more intuitive to duplicate the current (even unsaved) state of the request, instead of the saved stated though.

**Comment by willdean** [comment]
Until I read this issue, I have never known the duplicate function to work - it has always duplicated an empty query for me.  But as per @lauxjpn I have now realised that doing an explicit save first is what's needed.

I think this is a bit of an oddity - if duplicate doesn't work on an unsaved/modified request, then could it refuse to do the duplication, rather than just duplicating something unexpected?

**Comment by martyscholes** [comment]
Will,

At the risk of sounding stupid, how did you save a request? I chose the
Save option from the menu, but sometimes that did not seem to work. Did you
click on something before saving it?

**Comment by willdean** [comment]
@martyscholes I'm glad it's not just me who feels a bit dim here!  I clicked the "save" button over next to the "Send" button - I looked in all the various menus for a 'save' option first and didn't find it.   I can't say if 'sometimes' it doesn't work, because all I know is that I've never seen 'duplicate' work before ever (I guess I've always been duplicating requests which I've only just created from new), and now I've clicked that 'save' button (which I don't think I've ever clicked before), the duplication seems to work.  

Clearly some stuff is just saved implicitly and some stuff needs saving explicitly.

**Comment by martyscholes** [comment]
Will, thanks for getting back to me. For clarity, I have attached some
images. The first image shows what my screen looks like right now.

[image: Screen Shot 2019-04-26 at 11.23.19 AM.png]

If I want to save, I can click on the orange yin/yang button.

[image: Screen Shot 2019-04-26 at 11.24.55 AM.png]

But I don't know what that saves. Does it save the item in the collection
to the left?

[image: Screen Shot 2019-04-26 at 11.25.46 AM.png]

Or, does it save something related to the tabs at the top?

[image: Screen Shot 2019-04-26 at 11.26.21 AM.png]

Notice that the tab says [CONFLICT], suggesting that something is
conflicting with something else, but I don't know where to look for the
conflict.

When I click on the save button, what actually gets saved?

**Comment by vschoener** [comment]
Same issue here, duplicate only duplicate an empty request :(

**Comment by martyscholes** [comment]
**Solved?**

I have learned more about this issue and think I know what is going on. From what I now understand (I could be wrong), the requests on the left side of the screen are collection requests. The tabs at the top are _instances_ of collection requests. When clicking on a request in a collection, a new _instance_ (aka tab at the top) is created. Initially, the text in the tab name is italicized, indicating that nothing has been done. When the request is sent or any changes are made, the tab name is in normal text, indicating that the tab is now decoupled from the request in the collection.

If the tab has an orange dot, then the request in the tab has changed from the request in the collection. This is where the danger lies and where I got into trouble. I was ignoring the tabs at the top. When you click on the orange Yin/Yang circle at the top, nothing happens, despite the mouse hover text being, "Sync your API requests across devices." Clicking this circle saves nothing. At all. I kept clicking on this circle thinking I was saving the requests in the collections. I was not. I was doing nothing by clicking the orange circle.

If you click the Save button on the right near the top of the screen (something I never bothered to do), you will save the request in the current tab back to the request in the collection on the left. Do this often. I have made Command-S (the Save shortcut) a regular part of my routine in Postman. I cannot understand why auto-save is not available nor why it is not mandatory. When you save either by the keyboard or the Save button, the changes in the tab are applied to the request in the collection. At this point, the collection request is identical to the tab, and therefore the orange dot is removed from the tab.

For reasons unknown, it is possible for the same collection request to spawn multiple tabs. When this happens, Postman becomes confused as to which tab has the correct changes and all of the tabs tied to a particular collection request will have [CONFLICT] in the name, In this situation you have two options. If none of the tabs have changes that you want to keep, exit every tab and click the option to discard changes. If you want to keep the changes in one of the tabs, exit _every other_ tab for the same request and discard changes. With the single remaining tab, save it and the changes will be reflected back to the collection request.

If you click on a collection request and then make changes to it, the changes are only within the tab, not the collection request. No matter how many times you click on the orange circle, the changes will not be saved. If at this point you Duplicate a collection request, the duplicate will be of the state before you made the changes to the tab. This is why the Duplicate functionality appears broken: the wrong version fo the request is being duplicated.

The bottom line is to follow some simple rules:
- Save (Command-S, Control-S, or click the Save button) after every change to a request in a tab.
- Remove every tab not in use.
- Click on the orange circle as often as you like, but it will do nothing.

The above is the result of my experience and trying to read the Postman documentation. Any of this message could be factually incorrect. I hope it helps the next person.

**Comment by numaanashraf** [comment]
@martyscholes Thanks for the detailed writeup. You are in-fact right about how the current functionality works and we understand why these might be confusing to begin with. Just to summarise the current behavior:

1. requests in the sidebar are 'saved' requests which are saved within a collection in your workspace and also synced to your account and any teammates who are members in the same workspace.
2. opening requests opens them into a tab which you can use to send the request or make changes to the request - these are not automatically saved and need to be explicitly saved based via 'Save' or 'Command+S'
3. When you duplicate a request from the sidebar, we duplicate the current saved state of the request - not the unsaved state that might be open in a tab.

**Comment by numaanashraf** [comment]
@martyscholes @oskar-fjeld We've passed the feedback to our internal teams and we'ld be looking at how to make these more intuitive in upcoming releases. Closing this issue for now.

/cc @vkaegis

**Comment by eason-shsf** [comment]
> @oskar-fjeld Were all the changes to the request saved before you had duplicated the request?

this solved my problem, a saved request could be duplicated

**Comment by eromoe** [comment]
Still doesn't work on chrome 78.0.3904.108

**Comment by Gundark** [comment]
I was having this problem today.  I tried to duplicate a request using the ... next to the request in the collection on the sidebar.  Nothing at all happened.  Eventually, after several attempts, I noticed it had duplicated the entire collection, even though I used ... next to a specific request.  Then I tried to delete the duplicated collection, and Postman locked up entirely.  I couldn't save requests or anything.  I tried closing it, but that didn't help.  I eventually had to sign out to get Postman working again.  After signing out, and starting Postman from scratch, I used ... on the left hand sidebar to duplicate a specific request (not the whole collection), and it worked as expected.  So it looks like the feedback from last June was taken into account (it did what seemed intuitive), but there are some bugs causing it to not work / lock up in some situations.

**Comment by huqd** [comment]
Here I am, I did not know I have to explicitly click save the request then the duplicate feature can work as expected. Just did a search before leave it for another tool and see this thread.

**Comment by ohkjames** [comment]
Yea duplicate doesn't work, it stays empty....

**Comment by vschoener** [comment]
@ohkjames You have to save before duplicated, it works that way

**Comment by lauxjpn** [comment]
> Duplicate did not work for me as described.
> After explicitly saving the request I wanted to duplicate, it worked as expected.
> I think it would be more intuitive to duplicate the current (even unsaved) state of the request, instead of the saved stated though.

> We've passed the feedback to our internal teams and we'ld be looking at how to make these more intuitive in upcoming releases. Closing this issue for now.

@numaanashraf What is the status of this issue? It has been closed, but it seems that nothing has been done about it. So at a first step, please *reopen* this issue. It should only be closed, after it has been resolved.

The many comments of people stumbling over the current behavior should be prove enough, that it is unexpected.

**Comment by MrTrick** [comment]
I ran into this same issue. Seems like a very common stumbling book!

I think two small enhancements would help a **lot**:
- Display some small visual hint in the sidebar if the request has unsaved modifications.
- If duplicating a request with unsaved modifications, show a choice dialog, something like:
   ```
   This request has been modified and not yet saved.
     [ Duplicate original request ]
     [ Duplicate modified request ]
     [ Save original request, then duplicate ]

   [x] Remember my choice and don't ask again
  ```

**Comment by doransmestad** [comment]
Wow, yep.  This got me as well.  I've been using Postman for years and just thought it was broken until I finally got annoyed enough to go and open a bug ticket.  Turns out, it's just unintuitive behavior.   +1 to the proposal by @MrTrick.  

@numaanashraf  Looks like this ticket was closed a year ago - is there a new feature enhancement ticket (or something) we should be commenting on instead of this one to help drive this forward?

**Comment by arlemi** [comment]
@doransmestad #2838 tracks improvements to the two different duplicate mechanism we have in the app. You may want to add your feedback there.

**Comment by iulg** [comment]
It still happens in Postman v7.27.1

**Comment by anqilu** [comment]
It still happens in Postman v7.30.1. When performing duplicate on a request, the parameters are not correctly duplicated.

**Comment by willdean** [comment]
@anqilu I don't think the developers consider it a bug, so it's not going to be fixed.   The "duplicate" button duplicates the SAVED query, not the one you're currently working on.    So just get used to clicking "SAVE" (up at the top-right) before you click "Duplicate".  

It's a bit weird, but then it is free...

**Comment by anqilu** [comment]
@willdean It works as a miracle. Thanks for the advice.
I didn't notice the save button as it is a little "grey" in both dark/light theme. Now I find that all my queries are unsaved ones.

**Comment by willdean** [comment]
@anqilu As far as I am aware, the only use for the Save button is "the button I press before using Duplicate".

**Comment by jiangying000** [comment]
Maybe you could add an **autosave** option

**Comment by lauxjpn** [comment]
> > Duplicate did not work for me as described.
> > After explicitly saving the request I wanted to duplicate, it worked as expected.
> > I think it would be more intuitive to duplicate the current (even unsaved) state of the request, instead of the saved stated though.
> 
> > We've passed the feedback to our internal teams and we'ld be looking at how to make these more intuitive in upcoming releases. Closing this issue for now.
> 
> @numaanashraf What is the status of this issue? It has been closed, but it seems that nothing has been done about it. So at a first step, please **reopen** this issue. It should only be closed, after it has been resolved.
> 
> The many comments of people stumbling over the current behavior should be prove enough, that it is unexpected.

I'll ask again: What is the status of this issue? Fixing this should be prioritized. Until it is fixed, it should be **reopened**.

**Comment by diego-ramos** [comment]
> > > Duplicate did not work for me as described.
> > > After explicitly saving the request I wanted to duplicate, it worked as expected.
> > > I think it would be more intuitive to duplicate the current (even unsaved) state of the request, instead of the saved stated though.
> > 
> > 
> > > We've passed the feedback to our internal teams and we'ld be looking at how to make these more intuitive in upcoming releases. Closing this issue for now.
> > 
> > 
> > @numaanashraf What is the status of this issue? It has been closed, but it seems that nothing has been done about it. So at a first step, please **reopen** this issue. It should only be closed, after it has been resolved.
> > The many comments of people stumbling over the current behavior should be prove enough, that it is unexpected.
> 
> I'll ask again: What is the status of this issue? Fixing this should be prioritized. Until it is fixed, it should be **reopened**.

I had the same issue ( duplicate a request only creates an empty one ), but after read this post I found the tabs ( didn't notice them before ) and that they can ( and needed to ) be saved; after you saved the tab you can duplicate your request and the new one will have the data that you had saved before on the original request.

**Comment by lauxjpn** [comment]
> > > Duplicate did not work for me as described.
> > > After explicitly saving the request I wanted to duplicate, it worked as expected.
> > > I think it would be more intuitive to duplicate the current (even unsaved) state of the request, instead of the saved stated though.
> > 
> > 
> > > We've passed the feedback to our internal teams and we'ld be looking at how to make these more intuitive in upcoming releases. Closing this issue for now.
> > 
> > 
> > @numaanashraf What is the status of this issue? It has been closed, but it seems that nothing has been done about it. So at a first step, please **reopen** this issue. It should only be closed, after it has been resolved.
> > The many comments of people stumbling over the current behavior should be prove enough, that it is unexpected.
> 
> I'll ask again: What is the status of this issue? Fixing this should be prioritized. Until it is fixed, it should be **reopened**.

@numaanashraf It has been **9 months** since I asked about the status of this issue for the first time. I then asked again 4 months ago, and have not gotten any response either. Is anybody still monitoring this issue? If not, it looks like I have to create a new issue about the fact that comments on wrongfully closed issues do not get any support by the team and reference this issue here in it.

**Comment by torevor** [comment]
> @numaanashraf It has been **9 months** since I asked about the status of this issue for the first time. I then asked again 4 months ago, and have not gotten any response either. Is anybody still monitoring this issue? If not, it looks like I have to create a new issue about the fact that comments on wrongfully closed issues do not get any support by the team and reference this issue here in it.

Thanks to @martyscholes I at least now have a clue what is going on, and just noticed that yes indeed, I had numerous tabs open in Postman...thanks to the interesting colors and UI desig, I hadn't a clue they were open.  So after cleaning up the tabs, I now realize that I had somehow never saved these back into the collection, even though I had totally shut down Postman and reopened many times and was able to rerun them.

I have seen a lot of bad UI design choices in my lifetime, but I think this one is now my new leader.  Just unbelievable.

**Comment by Cap-JaTu** [comment]
This duplicate-feature which won't duplicate is really weird. Take it out, rename it to reflect the operation better or fix the damn thing!

**Comment by vschoener** [comment]
> This duplicate-feature which won't duplicate is really weird. Take it out, rename it to reflect the operation better or fix the damn thing!
It works, but you have to `save` before doing it...

An autosave feature would be better and more intuitive and would avoid having this issue anymore, wdyt ?

**Comment by Cap-JaTu** [comment]
> > This duplicate-feature which won't duplicate is really weird. Take it out, rename it to reflect the operation better or fix the damn thing!
>
> It works, but you have to `save` before doing it...
> An autosave feature would be better and more intuitive and would avoid having this issue anymore, wdyt ?

Yeah. I figured the save-part. It was painful, included cursing and wasn't pretty to watch.

On a computer, telling it to duplicate something makes a duplicate of the original thing. This can easily be achieved as a RAM-operation. Making a copy of a file is a completely separate thing.

With any object-oriented programming language, cloning an object is trivial. Doing a copy of a file on a filesystem requires couple lines more. Bottom line: **duplicate object** and **copy file** are different operations.

**Comment by y-usuzumi** [comment]
on mac, 9.14.11. I saved my request, duplicated it, and I got an empty request.

**Comment by coditva** [comment]
@y-usuzumi Can you open a new ticket with some screenshots/screen recording and logs from Developer Tools Console? It would help us debug this issue faster.
